pkg/controller/datavolume-controller.go

```diff
@@ -2601,3 +2638,236 @@ func GetRequiredSpace(filesystemOverhead float64, requestedSpace int64) int64 {
 func newLongTermCloneTokenGenerator(key *rsa.PrivateKey) token.Generator {
 	return token.NewGenerator(common.ExtendedCloneTokenIssuer, key, 10*365*24*time.Hour)
 }
+
+// detectCloneSize obtains and assigns the original PVC's size when cloning using an empty storage value
+func (r *DatavolumeReconciler) detectCloneSize(
+	dv *cdiv1.DataVolume,
+	pvcSpec *corev1.PersistentVolumeClaimSpec,
+	cloneType cloneStrategy) error {
+
+	var targetSize resource.Quantity
+	sourcePvc, err := r.findSourcePvc(dv)
+	if err != nil {
+		return err
+	}
+	sourceSize := sourcePvc.Status.Capacity.Storage()
+
+	// Due to possible filesystem overhead complications when cloning
+	// using host-assisted strategy, we create a pod that automatically
+	// collects the size of the original virtual image with 'qemu-img'.
+	// If another strategy is used or the original PVC's volume mode
+	// is "block", we simply extract the value from the original PVC's spec.
+	if cloneType == HostAssistedClone && getVolumeMode(sourcePvc) == corev1.PersistentVolumeFilesystem {
+		// If available, we first try to get the virtual size from previous iterations
+		imgSize, available := getSizeFromAnnotations(sourcePvc)
+		if !available {
+			imgSize, err = r.getSizeFromPod(sourcePvc, dv)
+			if err != nil {
+				return err
+			}
+		}
+		// Allow the clone-controller to skip the size comparison requirement
+		// if the source's size ends up being larger due to overhead differences
+		isSrcGreater := sourceSize.CmpInt64(imgSize) == 1
+		if isSrcGreater == true {
+			dv.Annotations[AnnPermissiveClone] = "true"
+		}
+		// If needed, inflate size with filesystem overhead
+		targetSize, err = inflateSizeWithOverhead(r.client, imgSize, pvcSpec)
+		if err != nil {
+			return err
+		}
+	} else {
+		targetSize = *sourceSize
+	}
+
+	pvcSpec.Resources.Requests[corev1.ResourceStorage] = targetSize
+	return nil
+}
+
+// getSizeFromAnnotations checks the source PVC's annotations and returns the requested size if it has already been obtained
+func getSizeFromAnnotations(sourcePvc *corev1.PersistentVolumeClaim) (int64, bool) {
+	virtualImageSize, available := sourcePvc.Annotations[AnnVirtualImageSize]
+	if available == true {
+		sourceCapacity, available := sourcePvc.Annotations[AnnSourceCapacity]
+		currCapacity := sourcePvc.Status.Capacity
+		// Checks if the original PVC's capacity has changed
+		if available && currCapacity.Storage().Cmp(resource.MustParse(sourceCapacity)) == 0 {
+			// Parse the raw string containing the image size into a 64-bit int
+			imgSizeInt, _ := strconv.ParseInt(virtualImageSize, 10, 64)
+			return imgSizeInt, true
+		}
+	}
+
+	return 0, false
+}
+
+// getSizeFromPod attempts to get the image size from a pod that directly obtains said value from the source PVC
+func (r *DatavolumeReconciler) getSizeFromPod(sourcePvc *corev1.PersistentVolumeClaim, dv *cdiv1.DataVolume) (int64, error) {
```

Comment:
I think it will be much cleaner if you remove the `getSizeFromPod()` and `datavolumeController.Watch(&source.Kind{Type: &corev1.Pod{}}`, and the containerized app can simply annotate the DV with the virtual size. See [this](https://github.com/kubevirt/containerized-data-importer/blob/main/tools/cdi-source-update-poller/main.go#L75-L76) for example.

Reply:
The question is always what is better. In the current design the app doesn't know it is inside a k8s cluster and can run independently, but at the cost of having the controller that manages it be more complex (it needs to create/start the pod, wait for it to finish, do some gymnastics to get the output, and then use the value) vs the pod now needs to know it is inside a k8s cluster, it needs to be passed a kubeconfig, you need to build a client, and the app needs to do the update to the datavolume which the controller can then act upon.

In this particular case IMO the extra complexity in the controller is better than teaching the app it is inside a k8s cluster, but it is a totally subjective opinion